### PR TITLE
Added ability to filter globally

### LINKED ISSUE #315: Be able to filter sensitive information
Sensitive information like creditcard information or passwords is likely to not want to be logged for security reasons.

Modifying and forking reporters for each and every single case isn't practical (once for file logging, for console, for udp, etc)

How this is to be done can be discussed. Perhaps a callback filter function, or a plugin for good itself.

See https://www.npmjs.com/package/blur  for reference

### PR DESCRIPTION
**Pull request by arb**
Closes #315.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -38,6 +38,14 @@ set `options` to an object with the following optional settings:
         - `key` - one of the supported good events or any of the `extensions` events that this reporter should listen for
         - `value` - a single string or an array of strings to filter incoming events. "\*" indicates no filtering. `null` and `undefined` are assumed to be "\*"
     - `config` - an implementation specific configuration value used to instantiate the reporter
+- `[filter]` - an object with the following keys:
```
Comment (danielb2) [inline README.md]: should this have brackets?

File: lib/utils.js [hunk 2]
```diff
@@ -67,15 +67,46 @@ exports.GreatError.prototype.toJSON = function () {
...
 
     if (this.constructor !== exports.GreatResponse) {
-        return new exports.GreatResponse(request, options);
+        return new exports.GreatResponse(request, options, filterRules);
     }
 
     var req = request.raw.req;
     var res = request.raw.res;
 
+    var replacer = function (match, group1) {
```
Comment (danielb2) [inline lib/utils.js]: call this variable `group` instead since there seems to be no `group2` ?

File: lib/utils.js [hunk 3]
```diff
@@ -67,15 +67,46 @@ exports.GreatError.prototype.toJSON = function () {
...
+
+        return (new Array(group1.length + 1).join('X'));
+    };
+
+    var applyFilter = function (data) {
+
+        for (var key in data) {
```
Comment (danielb2) [inline lib/utils.js]: hasOwnProperty check?

File: README.md (hunk 1 above)
Comment (arb) [inline README.md]: It's optional, so yeah I think brackets are good.

File: lib/utils.js [hunk 4]
```diff
@@ -67,15 +67,46 @@ exports.GreatError.prototype.toJSON = function () {
...
+            if (typeof data[key] === 'object') {
+                return applyFilter(data[key]);
+            }
+
+            // there is a filer for this key, so we are going to update the data
```
Comment (danielb2) [inline lib/utils.js]: There should be capitalized, and typo on `filer` -> `filter`

File: lib/utils.js [hunk 5]
```diff
@@ -67,15 +67,46 @@ exports.GreatError.prototype.toJSON = function () {
...
+            if (filterRules[key]) {
+                var filter = filterRules[key].toLowerCase();
+
+                if (filter === 'censor') {
+                    data[key] = ('' + data[key]).replace(/./gi, 'X');
```
Comment (danielb2) [inline lib/utils.js]: Don't need the `i` since `.` since you don't need to worry about case-insensitive

File: README.md [hunk 6]
```diff
@@ -38,6 +38,14 @@ set `options` to an object with the following optional settings:
...
+    - `key` - the key of the data property to change
+    - `value` - a string that can be one of the following:
+        - "censor" - replace the text with "X"s
+        - "remove" - `delete`s the value
+        - a valid regular express string. Only supports a single group. Ex: `"(\\d{4})$"` will replace the last four digits with "X"s. Take extra care when creating this string. You will need to make sure that the resultant RegExp object is what you need.
+
+    `filter` can be used to remove potentially sensitive information (credit card numbers, social security numbers, etc.) from the log payloads before they are sent out to reporters. This setting only impacts `response` events and only if payloads are included via `requestPayload` and `responsePayload`. `filter` is intended to impact the reporting of ALL downstream reporters. If you want filtering in only one, you will need to create a customized reporter.
```
Comment (danielb2) [inline README.md]: Should probably mention that this is recursive on the data. So the user has to be careful with the data to be manipulated/removed

File: lib/utils.js [hunk 7]
```diff
@@ -67,15 +67,46 @@ exports.GreatError.prototype.toJSON = function () {
...
+    var replacer = function (match, group1) {
+
+        return (new Array(group1.length + 1).join('X'));
+    };
+
+    var applyFilter = function (data) {
+
+        for (var key in data) {
+            if (typeof data[key] === 'object') {
+                return applyFilter(data[key]);
```
Comment (danielb2) [inline lib/utils.js]: What if the object is an array?

**Pull request merged**